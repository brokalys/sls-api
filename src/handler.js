import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
  ApolloServerPluginUsageReporting,
  ApolloServerPluginUsageReportingDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';
const isTestMode = process.env.NODE_ENV === 'test';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (isDevMode && !isTestMode) ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (!isTestMode) {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    !isDevMode && !isTestMode
      ? ApolloServerPluginUsageReporting({
          sendVariableValues: {
            exceptNames: ['email', 'unsubscribe_key'],
          },
          logger: {
            debug: () => {},
            info: () => {},
            warn: console.warn,
            error: console.error,
          },
          generateClientInfo: ({ context, request }) => {
            const headers = request.http && request.http.headers;
            const clientVersion =
              headers['apollographql-client-version'] || '1.0.0';

            if (!context.user) {
              return {
                clientName: 'Unauthorized',
                clientVersion,
              };
            }

            return {
              clientName: context.user.apiKey.name,
              clientVersion,
            };
          },
        })
      : ApolloServerPluginUsageReportingDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
