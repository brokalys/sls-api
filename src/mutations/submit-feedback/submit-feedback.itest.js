import { createTestClient } from 'apollo-server-testing';
import { server } from 'handler';
import Bugsnag from 'lib/bugsnag';

jest.mock('lib/bugsnag');

const mockInput = {
  type: 'bug-report',
  message: 'Hello World!',
  email: 'demo@example.com',
};

const { mutate } = createTestClient(server);

const MUTATION = `
mutation ChromeExtension_SubmitFeedback(
    $type: String!
    $message: String!
    $email: String
  ) {
    submitFeedback(
      type: $type
      message: $message
      email: $email
    )
  }
`;

describe('Mutation: submitFeedback', () => {
  test('successfully submits the feedback to Bugsnag', async () => {
    const response = await mutate({
      mutation: MUTATION,
      variables: {
        ...mockInput,
      },
    });

    expect(response).toMatchSnapshot();
    expect(Bugsnag.notify).toHaveBeenCalled();
  });
});
