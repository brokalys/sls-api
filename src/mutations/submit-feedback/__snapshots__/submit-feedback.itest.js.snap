// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation: submitFeedback successfully submits the feedback to Bugsnag 1`] = `
Object {
  "data": Object {
    "submitFeedback": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
